### PR TITLE
Delete heath data when nodes does not exist in node_lifecycle_controller

### PR DESCRIPTION
**Pull request by Garrybest**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

In kube-controller-manager, nodeHealthMap in node_lifecycle_controller.go does not have delete function. Memory leak may happen where nodes are created and deleted frequently, e.g., cluster composed of idle cloud cvm. When a node is deleted, its health data must be deleted from nodeHealthMap.

#### Which issue(s) this PR fixes:

Fixes #102630

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by Garrybest** [comment]
Sorry for waiting this late, add test for #102630 which cannot be reopened due to force push. @wzshiming

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 1]
```diff
@@ -52,6 +52,7 @@ import (
 	"k8s.io/client-go/util/workqueue"
 	nodetopology "k8s.io/component-helpers/node/topology"
 	kubeletapis "k8s.io/kubelet/pkg/apis"
+
```
Comment (wzshiming) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:55]:
```suggestion
```

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 2]
```diff
@@ -41,13 +42,14 @@ import (
 	"k8s.io/client-go/kubernetes/fake"
 	testcore "k8s.io/client-go/testing"
 	kubeletapis "k8s.io/kubelet/pkg/apis"
+	"k8s.io/utils/pointer"
```
Comment (wzshiming) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:45]: Move to the bottom of the `import`

**Review by wzshiming — CHANGES_REQUESTED** [review]
https://github.com/kubernetes/kubernetes/blob/aa66cec6fa6e6804669363d54dc8971a6cb59b79/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L189-L195

Since it can be deleted, a nil pointer check needs to be added here, and all callers of the method need to check the return result.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 3]
```diff
@@ -838,6 +846,10 @@ func (nc *Controller) monitorNodeHealth(ctx context.Context) error {
 			name := node.Name
 			node, err = nc.kubeClient.CoreV1().Nodes().Get(ctx, name, metav1.GetOptions{})
 			if err != nil {
+				// If the node no longer exists, delete the corresponding health data from nodeHealthMap.
```
Comment (wzshiming) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go:849]: I don't think it's necessary here, just remove it uniformly in the infomer's deletion callback function.

**Comment by Garrybest** [comment]
> Since it can be deleted, a nil pointer check needs to be added here, and all callers of the method need to check the return result.

Done. Note that nil pointer check is in function `func (n *nodeHealthData) deepCopy() *nodeHealthData`. We only need to check the callers which is in `handleDisruption`.

**Review by wzshiming — CHANGES_REQUESTED** [review]
https://github.com/kubernetes/kubernetes/blob/2aee0f2c3e91c86f5b266032331da09cb8eb8ffc/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1010

It seems that the check is missing here.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 4]
```diff
@@ -3840,3 +3841,134 @@ func Test_isNodeExcludedFromDisruptionChecks(t *testing.T) {
...
+		{
+			name:   "delete node0",
+			delete: nodes[0],
+			exists: nodes[1:],
+		},
+		{
+			name:   "delete node1",
+			delete: nodes[1],
+			exists: nodes[:1],
+		},
```
Comment (wzshiming) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:3897]:
```suggestion
		{
			name:   "delete node0",
			nodes:  nodes,
			delete: []string{"node0"},
			exists: []string{"node1"},
		},
		{
			name:   "delete node1",
			nodes:  nodes,
			delete: []string{"node1"},
			exists: []string{"node0"},
		},
```

**Comment by Garrybest** [comment]
> https://github.com/kubernetes/kubernetes/blob/2aee0f2c3e91c86f5b266032331da09cb8eb8ffc/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1010
> 
> It seems that the check is missing here.

I'm afraid not, see [here](https://github.com/kubernetes/kubernetes/blob/2aee0f2c3e91c86f5b266032331da09cb8eb8ffc/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1064-L1070).  The `nodeHealth` will be initialized if it is nil.

**Review by logicalhan — COMMENTED** [review]
/lgtm

**Comment by wzshiming** [comment]
It seems that the code is frozen today

**Comment by Garrybest** [comment]
Could you please take a look? @bowei

**Comment by Garrybest** [comment]
cc @bowei

**Pull request closed without merging**